internal/datacoord/garbage_collector.go

```diff
@@ -158,8 +163,28 @@ func (gc *garbageCollector) scan() {
 }
 
 func (gc *garbageCollector) clearEtcd() {
+	req := &internalpb.GetImportFailedSegmentIDsRequest{
+		Base: &commonpb.MsgBase{
+			MsgType: commonpb.MsgType_GetImportFailedSegmentIDs,
+		},
+	}
+	ctx := context.Background()
+	rsp, err := gc.rootCoordClient.GetImportFailedSegmentIDs(ctx, req)
+	if err != nil {
+		log.Error("getImportFailedSegmentIDs failed", zap.String("error", err.Error()))
+	}
+	failedSegmentIDs := rsp.SegmentIDs
+	failedSegmentIDsSet := make(map[int64]struct{})
+	for _, failedSegmentID := range failedSegmentIDs {
+		failedSegmentIDsSet[failedSegmentID] = struct{}{}
+	}
+
 	drops := gc.meta.SelectSegments(func(segment *SegmentInfo) bool {
-		return segment.GetState() == commonpb.SegmentState_Dropped && !gc.segRefer.HasSegmentLock(segment.ID)
+		_, failedSegmentFlag := failedSegmentIDsSet[segment.ID]
+		if failedSegmentFlag {
```

Comment:
This seems duplicated as the flag is checked in at line #187

Reply:
I will first set the segment state into dropped to make it easier to clear in the next time if the sevice is crashed.

Reply:
That makes sense. Then how about just remove the flag check in the next line?

Reply:
ok